tests/common/include/tests/common/tools/create_toy_geometry.hpp

```diff
@@ -9,432 +9,344 @@
-    rectangles.insert(rectangles.end(), l2_masks.begin(), l2_masks.end());
-
-    // Return all geometry containers
-    using geometry_t = toy_geometry<volume_type, surface>;
-
-    geometry_t geo(resource);
-    geo.add_volumes(std::move(volumes));
-    geo.add_objects(std::move(surfaces));
-
-    // First, put data into the detector interface
-    /*mask_store<dtuple, dvector, discs, cylinders, rectangles> masks;
-    // populate mask store
-    masks.add_masks(discs);
-    masks.add_masks(cylinders);
-    masks.add_masks(rectangles);*/
-    auto masks = std::make_tuple<disc_container, cylinder_container,
-                                 rectangle_container>(
-        std::move(discs), std::move(cylinders), std::move(rectangles));
-
-    auto d = toy_detector<decltype(geo), decltype(masks),
-                          decltype(transforms)::value_type, toy_grid>(
-        std::move(geo), std::move(masks), std::move(transforms));
-
-    return std::move(d);
+    // beampipe volue -- volume ID = 0
+    det.new_volume(
+        {0, beampipe_r, -1 * detector_half_z, detector_half_z, -M_PI, M_PI});
+
+    // first layer volume -- volume ID = 1
+    det.new_volume({beampipe_r, first_layer_outer_r, -1 * detector_half_z,
+                    detector_half_z, -M_PI, M_PI});
+
+    // gap volume -- volume ID = 2
+    det.new_volume({first_layer_outer_r, second_layer_inner_r,
+                    -1 * detector_half_z, detector_half_z, -M_PI, M_PI});
+
+    // second layer volume -- volume ID = 3
+    det.new_volume({second_layer_inner_r, second_layer_outer_r,
+                    -1 * detector_half_z, detector_half_z, -M_PI, M_PI});
+
+    auto& vol0 = det.volumes()[0];
+    auto& vol1 = det.volumes()[1];
+    auto& vol2 = det.volumes()[2];
+    auto& vol3 = det.volumes()[3];
+
+    /**
+     * Fill beampipe volume -- volume ID = 0
+     */
+
+    surface_container vol0_surfaces = {};
+    mask_container vol0_masks(resource);
+    transform_container vol0_transforms = {};
+
+    // disc portal at negative side
+    add_disc_portal(vol0.index(), ctx0, vol0_surfaces, vol0_masks,
+                    vol0_transforms, 0, beampipe_r, -detector_half_z,
+                    {leaving_world, inv_sf_finder});
+
+    // disc_portal at positive side
+    add_disc_portal(vol0.index(), ctx0, vol0_surfaces, vol0_masks,
+                    vol0_transforms, 0, beampipe_r, detector_half_z,
+                    {leaving_world, inv_sf_finder});
+
+    // beampipe cylinder potal -> linked to first layer (vol1)
+    add_cylinder_portal(vol0.index(), ctx0, vol0_surfaces, vol0_masks,
+                        vol0_transforms, beampipe_r, detector_half_z,
+                        {vol1.index(), inv_sf_finder});
+
+    // Add all objects to detector
+    det.add_objects(ctx0, vol0, vol0_surfaces, vol0_masks, vol0_transforms);
+
+    /**
+     * Fill the first layer volume -- volume ID = 1
+     */
+
+    surface_container vol1_surfaces = {};
```

Comment:
You can also just flush the previous containers

Reply:
only if there is a function to flush `mask_store`...